File: Cocoa/AppFrameworks/UIKit/CollectionViewManager/CollectionCell/YJUICollectionReusableView.m
//
//  YJUICollectionReusableView.m
//  YJUICollectionView
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 16/7/8.
//  Copyright © 2016年 YJCocoa. All rights reserved.
//

#import "YJUICollectionReusableView.h"
#import "YJUICollectionViewManager.h"
#import "YJDispatch.h"

#pragma mark - UICollectionReusableView (YJUICollectionView)
@implementation UICollectionReusableView (YJUICollectionView)

#pragma mark (+)
+ (YJUICollectionCellCreate)cellCreate {
    return YJUICollectionCellCreateDefault;
}

+ (id)cellObject {
    YJUICollectionCellObject *cellObject = [[YJUICollectionCellObject alloc] initWithCollectionViewCellClass:self.class];
    cellObject.createCell = [self cellCreate];
    return cellObject;
}

+ (id)cellObjectWithCellModel:(id<YJUICollectionCellModel>)cellModel {
    YJUICollectionCellObject *cellObject = [self cellObject];
    cellObject.cellModel = cellModel;
    return cellObject;
}

+ (CGSize)collectionViewManager:(YJUICollectionViewManager *)collectionViewManager viewForSupplementaryElementOfKind:(NSString *)kind referenceSizeForCellObject:(YJUICollectionCellObject *)cellObject {
    if (cellObject.createCell == YJUICollectionCellCreateDefault) { // 默认使用xib创建cell
        NSArray<UIView *> *array = [[NSBundle mainBundle] loadNibNamed:cellObject.cellName owner:nil options:nil];
        return array.firstObject.frame.size;
    }
    // 默认设置
    if ([UICollectionElementKindSectionHeader isEqualToString:kind]) {
        return collectionViewManager.delegateFlowLayoutManager.flowLayout.headerReferenceSize;
    }
    return collectionViewManager.delegateFlowLayoutManager.flowLayout.footerReferenceSize;
}

#pragma mark (-)
- (void)reloadDataWithCellObject:(YJUICollectionCellObject *)cellObject collectionViewManager:(YJUICollectionViewManager *)collectionViewManager {
    [self reloadDataSyncWithCellObject:cellObject collectionViewManager:collectionViewManager];
    __weakSelf
    dispatch_async_main(^{// UI加速
        [weakSelf reloadDataAsyncWithCellObject:cellObject collectionViewManager:collectionViewManager];
    });
}

- (void)reloadDataSyncWithCellObject:(YJUICollectionCellObject *)cellObject collectionViewManager:(YJUICollectionViewManager *)collectionViewManager {
}

- (void)reloadDataAsyncWithCellObject:(YJUICollectionCellObject *)cellObject collectionViewManager:(YJUICollectionViewManager *)collectionViewManager {
}

@end

#pragma mark YJUICollectionViewCell
@implementation YJUICollectionReusableView

- (void)reloadDataSyncWithCellObject:(YJUICollectionCellObject *)cellObject collectionViewManager:(nonnull YJUICollectionViewManager *)collectionViewManager {
    _cellObject = cellObject;
    _collectionViewManager = collectionViewManager;
}

@end

File: YJCocoa/Classes/AppFrameworks/UIKit/CollectionView/YJUICollectionReusableView.swift
//
//  YJUICollectionReusableView.swift
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/5/23.
//  Copyright © 2016-现在 YJCocoa. All rights reserved.
//

import UIKit

@objc extension UICollectionReusableView {
    
    /// 获取 YJUICollectionCellObject
    public class func cellObject() -> YJUICollectionCellObject {
        return YJUICollectionCellObject(cellClass: self)
    }
    
    /// 获取 YJUICollectionCellObject 并自动填充模型
    public class func cellObject(withCellModel cellModel: Any?) -> YJUICollectionCellObject {
        let co = self.cellObject()
        co.cellModel = cellModel
        return co
    }
    
    /// 获取 cell 的显示size
    open class func collectionViewManager(_ collectionViewManager: YJUICollectionViewManager, referenceSizeFor kind: String, in cellObject: YJUICollectionCellObject) -> CGSize {
        switch kind {
        case UICollectionView.elementKindSectionHeader:
            return collectionViewManager.flowLayout.headerReferenceSize
        case UICollectionView.elementKindSectionFooter:
            return collectionViewManager.flowLayout.footerReferenceSize
        default:
            return CGSize.zero
        }
    }
    
    /// 刷新 UICollectionViewCell
    open func collectionViewManager(_ collectionViewManager: YJUICollectionViewManager, reloadWith cellObject: YJUICollectionCellObject) {}
    
}

open class YJUICollectionReusableView: UICollectionReusableView {
    
    private(set) var cellObject: YJUICollectionCellObject!
    private(set) var collectionViewManager: YJUICollectionViewManager!
    
    override open func collectionViewManager(_ collectionViewManager: YJUICollectionViewManager, reloadWith cellObject: YJUICollectionCellObject) {
        super.collectionViewManager(collectionViewManager, reloadWith: cellObject)
        self.collectionViewManager = collectionViewManager
        self.cellObject = cellObject
    }
    
}
